sage.py

```diff
@@ -571,393 +485,271 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-    print('----- All unique servcies -----')
-    print(set([item for sublist in al_services for item in sublist]))
+            # Cut episode sequence when a low-severity episode follows a high-severity episode
+            count = 0
+            mcats = [epi[2] for epi in episodes]
+            cuts = [i for i in range(len(episodes) - 1) if (len(str(mcats[i])) > len(str(mcats[i + 1])))]#(ep[i] > 100 and ep[i+1] < 10)]
+            if len(cuts) == 0:
+                attacks.append(attack + '-0')
+                subsequences.append(episodes)
+                continue
+
+            rest = (-1, -1)
+            for i in range(len(cuts)):
+                start = 0 if i == 0 else cuts[i - 1] + 1
+                end = cuts[i]
+                rest = (end + 1, len(episodes) - 1)
+                subsequence = episodes[start:end+1]
+                if len(subsequence) < 2:
+                    continue
+                attacks.append(attack + '-' + str(count))
+                subsequences.append(subsequence)
+                count += 1
+            subsequence = episodes[rest[0]:rest[1]+1]
+            if len(subsequence) < 2:
+                #print('discarding symbol ', [x[2] for x in al]) # TODO This one is not cool1
+                continue
+            attacks.append(attack + '-' + str(count))
+            subsequences.append(subsequence)
+    print('\n# sub-sequences', len(attacks))
+    return subsequences, attacks
+
+
+## 27 Aug 2020: Generating traces for FlexFringe
+def generate_traces(subsequences, datafile):
+    all_services = [[most_frequent(epi[6]) for epi in subseq] for subseq in subsequences]
+    print('----- All unique services -----')
+    print(set([service for services_subseq in all_services for service in services_subseq]))
     print('---- end ----- ')
 
-    count_lines = 0
-    count_cats = set()
+    num_traces = 0
+    unique_mcat_mserv = set()  # FlexFringe treats the (mcat,mserv) pairs as symbols of the alphabet
 
-    f = open(datafile, 'w')
-    lengths = []
-    lines = []
-    for i,episodes in enumerate(victims):
-        #print(episodes)
-        num_behav = keys[i].split('-')[-1]
-        #print(keys[i], num_behav)
+    episode_traces = []
+    for i, episodes in enumerate(subsequences):
         if len(episodes) < 3:
             continue
-        #lengths+= len(episodes)
-        count_lines += 1
-        mcats = [str(x[2]) for x in episodes]
-        num_servs = [len(set((x[6]))) for x in episodes]
-        max_servs = [most_frequent(x[6]) for x in episodes]
+        num_traces += 1
+        mcats = [x[2] for x in episodes]
+        num_services = [len(set((x[6]))) for x in episodes]
+        max_services = [most_frequent(x[6]) for x in episodes]
         stime = [x[0] for x in episodes]
-        #print(stime)
-        #print(' '.join(mcats))
-
-        #multi = [str(c)+":"+str(n)+","+str(s) for (c,s,n) in zip(mcats,max_servs,num_servs)] # multivariate case
```

Comment:
Is this multivariable case still relevant? I have tried running with this line and got `KeyError: '4'`  on the line `sev = rev_smallmapping[sym.split('|')[0]]` in `traverse` method. So either this line (and some unused variables above) can be removed if this is not relevant anymore, or it has to be fixed so that it runs (although, I cannot say what the fix will include other than `mcat` and how to check the results, since I don't have the 'original' version.

Reply:
Yep, lets keep it in. The multivariate symbol is defined as: for each episode, the micro attack stage, the most frequently targeted service, the number of services in that episode.

Reply:
Alright, however I am not sure how it should be fixed. Right now it is `[str(c) + ":" + str(n) + "," + str(s) for (c, s, n) in zip(mcats, max_services, num_services)]`, however all the splits in attack graphs and lower are based on `|` symbol. Furthermore, a frequently occurring pattern is `rev_smallmapping[sym.split('|')[0]])`, which would break in case of multi. Could it then be `mcat|mserv,num_serv` or something like that? I think services are not really checked

Reply:
Actually, is the entire tuple a trace in FlexFringe or is it still `mcat|mserv`? Because if yes, then I am not sure if there will be any useful learning, and if no, then is it only needed for the states in the AGs?

![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/37b46dea-fbde-4f59-9ec8-89a439c5d4bf)
(This part is not doing anything for simply `mcat|mserv` but has to be modified depending on the format of the multivariable case